logfire/_internal/integrations/llm_providers/llm_provider.py

```diff
@@ -0,0 +1,162 @@
+from __future__ import annotations
+
+from contextlib import contextmanager
+from typing import TYPE_CHECKING, Any, AsyncIterator, Callable, ContextManager, Iterator
+
+from opentelemetry import context
+
+from ...constants import ONE_SECOND_IN_NANOSECONDS
+
+if TYPE_CHECKING:
+    from ...main import Logfire, LogfireSpan
+    from .types import EndpointConfig
+
+
+__all__ = ('instrument_llm_provider',)
+
+
+def instrument_llm_provider(
+    logfire: Logfire,
+    client: Any,
+    suppress_otel: bool,
+    scope_suffix: str,
+    get_endpoint_config_fn: Callable[[Any], EndpointConfig],
+    on_response_fn: Callable[[Any, LogfireSpan], Any],
+    is_async_client_fn: Callable[[Any], bool],
+) -> ContextManager[None]:
+    """Instruments the provided `client` with `logfire`."""
+    logfire_llm = logfire.with_settings(custom_scope_suffix=scope_suffix.lower(), tags=['llm'])
+
+    client._is_instrumented_by_logfire = True
+    client._original_request_method = original_request_method = client._request
+
+    is_async = is_async_client_fn(client)
+
+    def _instrumentation_setup(**kwargs: Any) -> Any:
+        if context.get_value('suppress_instrumentation'):
+            return None, None, kwargs
+
+        options = kwargs['options']
+        try:
+            message_template, span_data, content_from_stream = get_endpoint_config_fn(options)
+        except ValueError as exc:
+            logfire_llm.warn(
+                'Unable to instrument {suffix} API call: {error}', suffix=scope_suffix, error=str(exc), kwargs=kwargs
+            )
+            return None, None, kwargs
+
+        span_data['async'] = is_async
+        stream = kwargs['stream']
+
+        if stream and content_from_stream:
+            stream_cls = kwargs['stream_cls']
+            assert stream_cls is not None, 'Expected `stream_cls` when streaming'
+
+            class LogfireInstrumentedStream(stream_cls):
+                def __stream__(self) -> Iterator[Any]:
+                    with record_streaming(logfire_llm, span_data, content_from_stream) as record_chunk:
+                        for chunk in super().__stream__():  # type: ignore
+                            record_chunk(chunk)
+                            yield chunk
+
+            class LogfireInstrumentedAsyncStream(stream_cls):
+                async def __stream__(self) -> AsyncIterator[Any]:
+                    with record_streaming(logfire_llm, span_data, content_from_stream) as record_chunk:
+                        async for chunk in super().__stream__():  # type: ignore
+                            record_chunk(chunk)
+                            yield chunk
+
+            if is_async:
+                kwargs['stream_cls'] = LogfireInstrumentedAsyncStream
+            else:
+                kwargs['stream_cls'] = LogfireInstrumentedStream
+
+        return message_template, span_data, kwargs
+
+    def instrumented_llm_request_sync(**kwargs: Any) -> Any:
```

Comment:
For some reason this drops the second span containing the final streamed results, which breaks the streaming tests. Any ideas why this might be the case?

Reply:
Couldn't figure out why this was failing for stream. I think it's clean enough (at least for now) to resolve this and we can maybe address this in any future integration PRs (e.g. Groq)?